pkg/pod/status.go

```diff
@@ -353,59 +353,20 @@ func getWaitingMessage(pod *corev1.Pod) string {
 // sortTaskRunStepOrder sorts the StepStates in the same order as the original
 // TaskSpec steps.
 func sortTaskRunStepOrder(taskRunSteps []v1alpha1.StepState, taskSpecSteps []v1alpha1.Step) []v1alpha1.StepState {
-	trt := &stepStateSorter{
-		taskRunSteps: taskRunSteps,
-	}
-	trt.mapForSort = trt.constructTaskStepsSorter(taskSpecSteps)
-	sort.Sort(trt)
-	return trt.taskRunSteps
-}
-
-// stepStateSorter implements a sorting mechanism to align the order of the steps in TaskRun
-// with the spec steps in Task.
-type stepStateSorter struct {
-	taskRunSteps []v1alpha1.StepState
-	mapForSort   map[string]int
-}
-
-// constructTaskStepsSorter constructs a map matching the names of
-// the steps to their indices for a task.
-func (trt *stepStateSorter) constructTaskStepsSorter(taskSpecSteps []v1alpha1.Step) map[string]int {
-	sorter := make(map[string]int)
-	for index, step := range taskSpecSteps {
-		sorter[step.Name] = index
-	}
-	return sorter
-}
-
-// changeIndex sorts the steps of the task run, based on the
-// order of the steps in the task. Instead of changing the element with the one next to it,
-// we directly swap it with the desired index.
-func (trt *stepStateSorter) changeIndex(index int) {
-	// Check if the current index is equal to the desired index. If they are equal, do not swap; if they
-	// are not equal, swap index j with the desired index.
-	desiredIndex, exist := trt.mapForSort[trt.taskRunSteps[index].Name]
-	if exist && index != desiredIndex {
-		trt.taskRunSteps[desiredIndex], trt.taskRunSteps[index] = trt.taskRunSteps[index], trt.taskRunSteps[desiredIndex]
-	}
-}
-
-func (trt *stepStateSorter) Len() int { return len(trt.taskRunSteps) }
-
-func (trt *stepStateSorter) Swap(i, j int) {
-	trt.changeIndex(j)
-	// The index j is unable to reach the last index.
-	// When i reaches the end of the array, we need to check whether the last one needs a swap.
-	if i == trt.Len()-1 {
-		trt.changeIndex(i)
-	}
-}
-
-func (trt *stepStateSorter) Less(i, j int) bool {
-	// Since the logic is complicated, we move it into the Swap function to decide whether
-	// and how to change the index. We set it to true here in order to iterate all the
-	// elements of the array in the Swap function.
-	return true
+	sort.Slice(taskRunSteps, func(i, j int) bool {
+		iStep := taskRunSteps[i]
+		jStep := taskRunSteps[j]
+		for _, specStep := range taskSpecSteps {
+			if iStep.Name == specStep.Name {
+				return true
+			}
+			if jStep.Name == specStep.Name {
+				return false
+			}
+		}
+		return false
+	})
+	return taskRunSteps
```

Comment:
is this part of fixing the issue or is it a separate improvement? if the latter, maybe a separate PR makes sense?

Reply:
So, the panic originates here.  I have two thoughts related to this :-
1. we shouldn't really be leaving a sort func lying around that panics in the face of a length mismatch.
2. this sorting code felt very complicated when compared to its intended purpose. I spent a good amount of time just trying to puzzle out what it was trying to achieve when the out-of-bounds error occurred. I would guess that the number of steps in a task is always going to be relatively small and I'm not convinced that the performance optimization this func was implementing was really worthwhile (even if the number of steps went into the hundreds or thousands... I don't think this would have changed the runtime really dramatically).

I guess we could approach this a couple of different ways (that I can think of, happy to go a different way if there are more I don't mention):
- as done here, sort the statuses and push any that don't appear in taskRunSteps to the back of the list.  I should probably also add a comment describing this behaviour.
- check if len(taskRunSteps) and len(taskSpecSteps) are equal.  If not, return the list unsorted.  Leave the stepStateSorter in place. comment about the constraint.
- document the panic with a comment.
- leave alone for now and revisit in a separate PR.

I'm kinda happy with the change as it stands (though it needs a comment) but I also don't mind if you think we should tackle this in a separate PR or a totally different way.  WDYT?

Reply:
i've added a comment describing the new sort behaviour.  still happy to take this change out or modify if preferred.

Reply:
Nit (and note): `sort.Slice` uses reflection (whereas `sort.Sort` doesn't, rely on the `interface` contract). This means, it's gonna be slower — might be a premature optimization thinking that though :stuck_out_tongue:

Reply:
I've put the original sort func back.  Don't want to hold up this PR if that change is controversial.